### PR TITLE
CATROID-162 Fix issues with renaming during upload

### PR DESCRIPTION
**Pull request by thmq**

### CODE REVIEW

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java [hunk 1]
```diff
@@ -126,104 +216,196 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+			project = ProjectManager.getInstance().getCurrentProject();
 		}
 
-		Project project = ProjectManager.getInstance().getCurrentProject();
+		project.setDescription(description);
+		project.setDeviceData(this);
+
+		File projectDir = new File(DEFAULT_ROOT_DIRECTORY, project.getName());
 
-		if (Utils.isDefaultProject(project, this)) {
```
Comment (dpedratscher) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java]: are we losing this functionality at this point? is there a check somewhere else if it's the default project?

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java [hunk 2]
```diff
@@ -126,104 +216,196 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+						setNextButtonEnabled(true);
+					}
+				})
+				.setCancelable(false)
+				.show();
 	}
 
-	private void onNextButtonClick() {
-		String name = nameInputLayout.getEditText().getText().toString().trim();
+	private void showProgressDialogAndUploadProject() {
```
Comment (dpedratscher) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java]: this function is quite long

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java [hunk 3]
```diff
@@ -126,104 +216,196 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
-		}
+		new AlertDialog.Builder(this)
+				.setTitle(getString(R.string.rating_dialog_title))
+				.setView(R.layout.dialog_rate_pocketcode)
+				.setPositiveButton(R.string.rating_dialog_rate_now, new DialogInterface.OnClickListener() {
+					@Override
+					public void onClick(DialogInterface dialog, int id) {
+						try {
+							startActivity(new Intent(Intent.ACTION_VIEW,
+									Uri.parse("market://details?id=" + getPackageName())).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK));
```
Comment (dpedratscher) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java]: could "market://details" also be defined as constant?

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java [hunk 4]
```diff
@@ -126,104 +216,196 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+		final int notificationId = StatusBarNotificationManager.getInstance()
+				.createUploadNotification(this, name);
+
+		uploadProgressDialog = new AlertDialog.Builder(this)
+				.setTitle(getString(R.string.upload_project_dialog_title))
+				.setView(R.layout.dialog_upload_project_progress)
+				.setPositiveButton(R.string.progress_upload_dialog_show_program, null)
+				.setNegativeButton(R.string.done, new DialogInterface.OnClickListener() {
+					@Override
+					public void onClick(DialogInterface dialog, int which) {
```
Comment (dpedratscher) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java]: here the variable "which" is again. down on 354 the name if the variable is "id". Would be clearer if they had the same name.

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java [hunk 5]
```diff
@@ -126,104 +216,196 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+				.setView(R.layout.dialog_rate_pocketcode)
+				.setPositiveButton(R.string.rating_dialog_rate_now, new DialogInterface.OnClickListener() {
+					@Override
+					public void onClick(DialogInterface dialog, int id) {
+						try {
+							startActivity(new Intent(Intent.ACTION_VIEW,
+									Uri.parse("market://details?id=" + getPackageName())).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK));
+						} catch (ActivityNotFoundException e) {
+							startActivity(new Intent(Intent.ACTION_VIEW,
+									Uri.parse("https://play.google.com/store/apps/details?id=" + getPackageName())));
```
Comment (dpedratscher) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java]: could this please be defined as a constant somewhere?

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java [hunk 6]
```diff
@@ -260,42 +436,39 @@ public void onTagsReceived(List<String> tags) {
...
-			if (FileMetaDataExtractor.getProjectNames(DEFAULT_ROOT_DIRECTORY).contains(input)) {
-				return getString(R.string.name_already_exists);
+		@Override
+		protected Void doInBackground(Void... voids) {
+			while (StatusBarNotificationManager.getInstance().getProgressPercent() != 100) {
+				try {
+					Thread.sleep(100);
+				} catch (InterruptedException e) {
+					Log.e(TAG, "Well, that's awkward.");
+				}
```
Comment (dpedratscher) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java]: why u no funny? :D

File: catroid/src/main/java/org/catrobat/catroid/io/ZipArchiver.java [hunk 7]
```diff
@@ -43,6 +43,7 @@
 	private static final int COMPRESSION_LEVEL = 0;
 
 	public void zip(File archive, File[] files) throws IOException {
+		archive.createNewFile();
```
[the remark this answers is not in the document]
Comment (LPeteR90) [reply catroid/src/main/java/org/catrobat/catroid/io/ZipArchiver.java]: createNewFile does not create a new file if a file already exists on the path... I know, not very expected behaviour, but hey, Java :)

File: catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectSaveTask.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+		weakListenerReference = new WeakReference<>(listener);
+	}
+
+	public static boolean task(Project project) {
+		return XstreamSerializer.getInstance().saveProject(project);
+	}
+
+	@Override
+	protected Boolean doInBackground(Project... projects) {
+		return task(projects[0]);
```
[the remark this answers is not in the document]
Comment (thmq) [reply catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectSaveTask.java]: because you can only save one project at a time, while you can import multiple projects within the same scope.

why don't we allow to save multiple projects at once you may ask, that's a good question, I'd reply. however, the `XstreamSerializer` is quite peculiar when it comes to loading and saving projects (i.e. we haven't yet refactored it) and the what the methods used for importing do on the other hand is clearer (i.e.we recently wrote them) so we are more confident that importing multiple projects is less likely to blow up than saving. besides I did not really see a use case where I'd want to save multiple projects at the same time, as our app only supports one loaded project (that can be modified) at a time anyway.

File: catroid/src/main/java/org/catrobat/catroid/ui/MainMenuActivity.java [hunk 9]
```diff
@@ -239,42 +274,13 @@ private void importProjectsFromExternalStorage() {
 				.setPositiveButton(R.string.import_dialog_move_btn, new DialogInterface.OnClickListener() {
 					@Override
 					public void onClick(DialogInterface dialog, int which) {
```
[the remark this answers is not in the document]
Comment (thmq) [reply catroid/src/main/java/org/catrobat/catroid/ui/MainMenuActivity.java]: [this](https://developer.android.com/reference/android/content/DialogInterface.OnClickListener.html#onClick(android.content.DialogInterface,%20int)) may come in handy here :wink:. 

**TL;DR** `which` button e.g. `DialogInterface.BUTTON_POSITIVE` was clicked.

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java [hunk 10]
```diff
@@ -73,10 +80,61 @@
...
 
-	private TextChangedListener textChangedListener = new TextChangedListener();
+	public static final int MAX_NUMBER_OF_TAGS_CHECKED = 3;
+	public static final String NUMBER_OF_UPLOADED_PROJECTS = "number_of_uploaded_projects";
+
+	private Project project;
+
+	private AlertDialog uploadProgressDialog;
+
+	private TextWatcher nameInputTextWatcher = new TextWatcher() {
```
[the remark this answers is not in the document]
Comment (thmq) [reply catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java]: **short answe**r: no
**long answer**: yes, but I could only add a `private` class at the end that contains all the overrides and then instantiate it here. so I could move the class definition somewhere else, but the member has to stay here.

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java (hunk 2 above)
Comment (thmq) [reply catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java]: yes, so what do want me to do about it? IMO it is clear what the snippet does (i.e. it get the `String name` from `nameInputLayout`.

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java (hunk 3 above)
Comment (thmq) [reply catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java]: yes, it could, but what's the benefit? it's never used anywhere else.

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java (hunk 4 above)
Comment (thmq) [reply catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java]: image

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java (hunk 1 above)
Comment (thmq) [reply catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java]: yes! you're absolutely right! I missed that one. thanks :wink:

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java (hunk 4 above)
Comment (dpedratscher) [reply catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java]: Hi, sorry I don't get it. Are these two different DialogInterfaces? Or are the varnames mixed?
ProjectUploadActivity, 288, 326 -- varname = which
ProjectUploadActivity, 237, 244, 339 -- varname = id

Sorry for this pedantic stuff again. I really spent a lot of time trying to find / understand a difference.

Comment (thmq) [reply catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java]: this is just android studio's auto-create. and TBH I don't usually check every auto-created varname with every name that was already in the code somewhere.

**Comment by ndrnour** [comment]
well done, what do you think about changing the title of the dialog in the attached screen-shot to something else, since the user has not anymore to choose up to 3 categories?
upload categories

File: catroid/src/main/java/org/catrobat/catroid/common/Constants.java [hunk 11]
```diff
@@ -111,6 +111,8 @@
 	public static final String CATROBAT_TOKEN_LOGIN_URL = BASE_URL_HTTPS + "tokenlogin?username=";
 	public static final String CATROBAT_DELETE_ACCOUNT_URL = BASE_URL_HTTPS + "profile/edit";
 
+	public static final String PLAY_STORE_PAGE_LINK = "https://play.google.com/store/apps/details?id=";
```
Comment (ndrnour) [inline catroid/src/main/java/org/catrobat/catroid/common/Constants.java]: I guess no need for this any more. right?

Comment (thmq) [reply catroid/src/main/java/org/catrobat/catroid/common/Constants.java]: I just added it. It's when you are asked to rate the app.

Comment (ndrnour) [reply catroid/src/main/java/org/catrobat/catroid/common/Constants.java]: actually I dont see any usage for it in your Pr, do you mean here `		Uri.parse("market://details?id=" + getPackageName())).addFlags(Intent.FLAG_ACTIVITY_NEW_TASK));`

**Comment by thmq** [comment]
apparently #3095 apparently broke something / or something has changed on jenkins, the same test that only failed on #3095 no crashes here as well.

**Comment by thmq** [comment]
> well done!
> what do you think about changing the title of the dialog in the attached screen-shot to something else, since the user has not anymore to choose up to 3 categories?
> upload categories

IMO this can also mean that you don't have to select anything, but I guess @wslany would have to decide about the string for this dialog :wink:

File: catroid/src/main/java/org/catrobat/catroid/common/Constants.java (hunk 11 above)
Comment (thmq) [reply catroid/src/main/java/org/catrobat/catroid/common/Constants.java]: [here](https://github.com/Catrobat/Catroid/pull/3104/files#diff-283bd7edee90877049fecac924c059cfR332) it is!

Comment (ndrnour) [reply catroid/src/main/java/org/catrobat/catroid/common/Constants.java]: My bad, I used "find in page" to look for the usage, it seems broken in my browser. *_#

**Comment by wslany** [comment]
> Re: Choose up to three categories
Was there any change here?

**Comment by thmq** [comment]
> > Re: Please choose up to three categories
> 
> Was there any change here?

well, IMO not, but @ndrnour mentioned that he thinks that previously you had to select at least one category for the next button to become enabled. (although I do not think this is true).

**Comment by wslany** [comment]
> well, IMO not, but @ndrnour mentioned that he thinks that previously you had to select at least one category for the next button to become enabled. (although I do not think this is true).

No, that was always allowed, and is correct as it is.

**Comment by wslany** [comment]
Ignoring the unrelated failures (need to be corrected independently).
